jsonrpc/ipc_all.go

```diff
@@ -0,0 +1,69 @@
+package jsonrpc
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+)
+
+const (
+	// On Linux, path is 108 bytes in size.
+	// http://man7.org/linux/man-pages/man7/unix.7.html
+	maxUnixPathBytes = 108
+	// fileModePerms represents the permission mode for directories, which is set to 751.
+	// This mode is stricter than the common default of 755 for directories.
+	fileModePerms = 0o751
+	// socketPerms represents the permission mode for socket files, set to 600. It allows
+	// read and write access for the owner only, ensuring that socket is accessible only to the intended user.
+	socketPerms = 0o600
+)
+
+var errPathTooLong = errors.New("path too long")
+
+// createListener creates a Unix domain socket at the given path and sets proper permissions.
+func createListener(endpoint string) (net.Listener, error) {
+	var (
+		l   net.Listener
+		err error
+	)
+	// path + terminator
+	if len(endpoint)+1 > maxUnixPathBytes {
+		return nil, errPathTooLong
+	}
+	// Try connecting first; if it works, then the socket is still live,
+	// so let's abort the creation of a new one.
+	c, err := net.Dial("unix", endpoint)
+	if err == nil {
+		c.Close()
+		return nil, fmt.Errorf("%v: address already in use", endpoint)
+	}
+	err = preparePath(endpoint)
+	if err != nil {
```

Comment:
```suggestion
	if err = preparePath(endpoint); err != nil {
```
Let's inline `preparePath` if it's only used here.